lib/galaxy_ext/metadata/set_metadata.py

```diff
@@ -70,8 +71,65 @@ def set_meta_with_tool_provided(dataset_instance, file_dict, set_meta_kwds, data
 
 
 def set_metadata():
-    # locate galaxy_root for loading datatypes
-    galaxy_root = os.path.abspath(os.path.join(os.path.dirname(__file__), os.pardir, os.pardir, os.pardir))
+    if len(sys.argv) == 1:
+        set_metadata_portable()
+    else:
+        set_metadata_legacy()
+
+
+def set_metadata_portable():
+    import galaxy.model
+    galaxy.model.metadata.MetadataTempFile.tmp_dir = tool_job_working_directory = os.path.abspath(os.getcwd())
+
+    metadata_params_path = os.path.join("metadata", "params.json")
+    with open(metadata_params_path, "r") as f:
+        metadata_params = json.load(f)
+
+    datatypes_config = metadata_params["datatypes_config"]
+    job_metadata = metadata_params["job_metadata"]
+    max_metadata_value_size = metadata_params.get("max_metadata_value_size") or 0
+    outputs = metadata_params["outputs"]
+
+    datatypes_registry = validate_and_load_datatypes_config(datatypes_config)
+    tool_provided_metadata = load_job_metadata(job_metadata)
+
+    def set_meta(new_dataset_instance, file_dict):
+        set_meta_with_tool_provided(new_dataset_instance, file_dict, set_meta_kwds, datatypes_registry, max_metadata_value_size)
+
+    for output_name, output_dict in outputs.items():
+        filename_in = os.path.join("metadata/metadata_in_%s" % output_name)
+        filename_kwds = os.path.join("metadata/metadata_kwds_%s" % output_name)
+        filename_out = os.path.join("metadata/metadata_out_%s" % output_name)
+        filename_results_code = os.path.join("metadata/metadata_results_%s" % output_name)
+        override_metadata = os.path.join("metadata/metadata_override_%s" % output_name)
+        dataset_filename_override = output_dict["filename_override"]
+
+        # Same block as below...
+        set_meta_kwds = stringify_dictionary_keys(json.load(open(filename_kwds)))  # load kwds; need to ensure our keywords are not unicode
+        try:
+            dataset = cPickle.load(open(filename_in, 'rb'))  # load DatasetInstance
```

Comment:
Unrelated comment: I think pickling might bite us if we really start doing portable metadata generation (plus it's a performance hit I think). We just need very little from the dataset instance, it just needs to be compatible for the set_meta methods in the dataype framework, right ? We could mock this out, which would also allow us to do metadata setting without the model / with a very minimal model.

Reply:
Once my history import/export rewrite branch that introduces the concept of a JSON model store merges with my remote metadata branch - we will be using json, not pickles for remote metadata.

Reply:
I guess this is a slightly different context though - still here I agree it would be good to do this. I’ll try to keep this in mind.